### PR TITLE
feat(broken upgrade path): [Internal Schedules] The color field should be in the category and not in the session

### PR DESCRIPTION
**Pull request by rollki**
So, this PR and changes to it change how the integration with colors in the calendar happens.

Now the color must be placed directly in the entity "activity" in **hex** format, after that all events on the calendar that are related to this activity will be automatically marked with a color.

That is, with this change, we will not have color control over individual events in the calendar, but only over events connected to one activity.

Screenshot 2024-04-02 at 16 22 40
Screenshot 2024-04-02 at 16 23 35
Screenshot 2024-04-02 at 16 23 49

**Steps to reproduce:**
1. Go to the content viewing page and filter by the essence of "activity" (/admin/content?title=&type=activity&status=All&langcode=All)
2. Edit one or more by adding a hex color to them
3. Go to the page with calendars
4. Events should automatically be displayed with the specified color for the activity

### CODE REVIEW

**Comment by rollki** [comment]
This PR contains changes from https://github.com/rollki/y_pef_schedule/pull/9

File: src/Controller/FullCalendarController.php [hunk 1]
```diff
@@ -418,22 +420,32 @@ public function getCategories(): JsonResponse {
...
     return new JsonResponse($categories);
   }
 
+  /**
+   * Get default color event if color not set in activity node
+   *
+   * @return string color event
+   */
+  public static function getDefaultColor() {
+    return '#3788d8';
```
Comment (podarok) [inline src/Controller/FullCalendarController.php]: should be a drupal configuration

File: y_pef_schedule.install [hunk 2]
```diff
@@ -27,16 +27,16 @@ function y_pef_schedule_uninstall(): void {
...
  */
 function _y_pef_schedule_create_fields(): void {
-  // Define and save field storage for 'Session color'.
+  // Define and save field storage for 'activity color'.
   FieldStorageConfig::create([
-    'field_name' => 'field_session_color',
+    'field_name' => 'field_activity_color',
     'entity_type' => 'node',
     'type' => 'string',
     'settings' => [
```
Comment (podarok) [inline y_pef_schedule.install]: Let's reuse colorapi for better CX, please

File: y_pef_schedule.install [hunk 3]
```diff
@@ -45,34 +45,34 @@ function _y_pef_schedule_create_fields(): void {
     'cardinality' => 1,
   ])->save();
 
-  // Define and save field configuration for 'Session color'.
+  // Define and save field configuration for 'activity color'.
   FieldConfig::create([
-    'field_name' => 'field_session_color',
+    'field_name' => 'field_activity_color',
```
Comment (podarok) [inline y_pef_schedule.install]: why not config?

Comment (rollki) [reply y_pef_schedule.install]: @david-storm 
I agree, we can probably use a simple exported config

File: src/Controller/FullCalendarController.php (hunk 1 above)
Comment (rollki) [reply src/Controller/FullCalendarController.php]: Moved.

File: y_pef_schedule.install (hunk 2 above)
Comment (rollki) [reply y_pef_schedule.install]: Done

File: y_pef_schedule.info.yml [hunk 4]
```diff
@@ -8,3 +8,4 @@ version: 1.0
 lifecycle: experimental
 dependencies:
   - drupal:openy_repeat
+  - colorapi
```
Comment (podarok) [inline y_pef_schedule.info.yml]:
```suggestion
  - drupal:colorapi
```

File: config/install/field.storage.node.field_activity_color.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+field_name: field_activity_color
+entity_type: node
+type: colorapi_color_field
+settings: {  }
+module: colorapi
+locked: false
+cardinality: 1
+translatable: true
+indexes: {  }
+persist_with_no_fields: false
```
Comment (podarok) [inline config/install/field.storage.node.field_activity_color.yml]:
```suggestion
persist_with_no_fields: true
```

for safer upgrade path

File: y_pef_schedule.install [hunk 6]
```diff
@@ -12,8 +12,8 @@ use Drupal\field\Entity\FieldConfig;
  * Implements hook_install().
  */
 function y_pef_schedule_install(): void {
-  // Create custom fields required by the module.
-  _y_pef_schedule_create_fields();
+  // Update display settings for the 'activity color' field.
+  _y_pef_schedule_update_displays_on_install();
```
Comment (podarok) [inline y_pef_schedule.install:16]: this is a major change by definition, you changed install process
So module will get major version bump after merge

File: y_pef_schedule.install [hunk 7]
```diff
@@ -73,7 +39,7 @@ function _y_pef_schedule_update_displays_on_install(): void {
   // Retrieve and update the form display for 'activity' nodes.
   $form_display = $entity_display_repository->getFormDisplay('node', 'activity');
   $form_display->setComponent('field_activity_color', [
-    'type' => 'string_textfield',
+    'type' => 'colorapi_color_widget',
```
Comment (podarok) [inline y_pef_schedule.install:42]: we need to ensure upgrade path is supported
What happens with existing installation?
Are we sure nobody uses this to date?

**Review by podarok — CHANGES_REQUESTED** [review]
Some suggestions and questions

File: y_pef_schedule.install (hunk 7 above)
Comment (podarok) [reply y_pef_schedule.install:42]: we are in rc phase, so I assume we are safe, but we would need to describe after rc6 we have broken upgrade path ( possibly )

File: y_pef_schedule.install (hunk 6 above)
Comment (podarok) [reply y_pef_schedule.install:16]: we are in rc phase, so I assume we are safe, but we would need to describe after rc6 we have broken upgrade path ( possibly )

**Comment by rollki** [comment]
Hi @podarok 
Thank you for your comments.
Precisely because we are in the RC phase, we did not do the upgrade path, since this module is not used yet, it is used only on one build (https://github.com/YCloudYUSA/y_pef_schedule/pull/10), which actually resets and updates the configuration, so even there it is not a problem. So do we really want to add it?

**Comment by podarok** [comment]
> Hi @podarok Thank you for your comments. Precisely because we are in the RC phase, we did not do the upgrade path, since this module is not used yet, it is used only on one build (#10), which actually resets and updates the configuration, so even there it is not a problem. So do we really want to add it?

we can skip upgrade path work based on a knowledge nobody using it to date

File: y_pef_schedule.install (hunk 3 above)
Comment (podarok) [reply y_pef_schedule.install]: Are you going to do it in thic PR or we can merge as is and do a follow-up? @rollki

**Comment by rollki** [comment]
@podarok 
Can we merge it then?

**Comment by podarok** [comment]
https://github.com/YCloudYUSA/y_pef_schedule/releases/tag/1.0.0-rc7

**Pull request merged**